drivers/SmartThings/zigbee-switch/src/zll-dimmer-bulb/ikea-xy-color-bulb/init.lua

```diff
@@ -64,38 +67,73 @@ local query_device = function(device)
 end
 
 local function set_color_handler(driver, device, cmd)
-  local hue = cmd.args.color.hue > 99 and 99 or cmd.args.color.hue
+  -- Cancel the hue/sat timer if it's running, since setColor includes both hue and saturation
+  local huesat_timer = device:get_field(HUESAT_TIMER)
+  if huesat_timer ~= nil then
+    device.thread:cancel_timer(huesat_timer)
+    device:set_field(HUESAT_TIMER, nil)
+  end
+
+  local hue = (cmd.args.color.hue ~= nil and cmd.args.color.hue > 99) and 99 or cmd.args.color.hue
   local sat = cmd.args.color.saturation
+
   local x, y, Y = utils.safe_hsv_to_xy(hue, sat)
   store_xyY_values(device, x, y, Y)
-  switch_defaults.on(driver,device,cmd)
+  switch_defaults.on(driver, device, cmd)
 
   device:send(ColorControl.commands.MoveToColor(device, x, y, 0x0000))
 
+  device:set_field(TARGET_HUE, nil)
+  device:set_field(TARGET_SAT, nil)
   device.thread:call_with_delay(2, query_device(device))
 end
 
-local function set_hue_handler(driver, device, cmd)
-  local sat = device:get_latest_state("main", capabilities.colorControl.ID, capabilities.colorControl.saturation.NAME)
-  local hue = cmd.args.hue > 99 and 99 or cmd.args.hue
-  local x, y, Y = utils.safe_hsv_to_xy(hue, sat)
-  store_xyY_values(device, x, y, Y)
-  switch_defaults.on(driver,device,cmd)
+local huesat_timer_callback = function(driver, device, cmd)
+  return function()
+    local hue = device:get_field(TARGET_HUE)
+    local sat = device:get_field(TARGET_SAT)
+    hue = hue ~= nil and hue or device:get_latest_state("main", capabilities.colorControl.ID, capabilities.colorControl.hue.NAME)
+    sat = sat ~= nil and sat or device:get_latest_state("main", capabilities.colorControl.ID, capabilities.colorControl.saturation.NAME)
+    cmd.args = {
+      color = {
+        hue = hue,
+        saturation = sat
+      }
+    }
+    set_color_handler(driver, device, cmd)
+  end
+end
 
-  device:send(ColorControl.commands.MoveToColor(device, x, y, 0x0000))
+local function set_hue_sat_helper(driver, device, cmd, hue, sat)
+  local huesat_timer = device:get_field(HUESAT_TIMER)
+  if huesat_timer ~= nil then
+    device.thread:cancel_timer(huesat_timer)
+    device:set_field(HUESAT_TIMER, nil)
+  end
+  if hue ~= nil and sat ~= nil then
+    cmd.args = {
+      color = {
+        hue = hue,
+        saturation = sat
+      }
+    }
+    set_color_handler(driver, device, cmd)
+  else
+    if hue ~= nil then
+      device:set_field(TARGET_HUE, hue)
+    elseif sat ~= nil then
+      device:set_field(TARGET_SAT, sat)
+    end
+    device:set_field(HUESAT_TIMER, device.thread:call_with_delay(0.2, huesat_timer_callback(driver, device, cmd)))
+  end
+end
 
-  device.thread:call_with_delay(2, query_device(device))
+local function set_hue_handler(driver, device, cmd)
+  set_hue_sat_helper(driver, device, cmd, cmd.args.hue, device:get_field(TARGET_SAT))
```

Comment:
So it looks like this same cmd argument table is passed through the whole time.  It seems a bit questionable to be passing around a `setHue` command with non-standard args, and passing it into a `setColor` command handler.  In this case everything _works_ but it's a bit fragile.

Reply:
I guess since this is a specific sub-driver I'm OK with it.  I was originally worried about accidental usage of this not maintaining the assumptions, but given it's for a single driver it seems fine.